compiler/plugins/target/AMD-AIE/aie/AMDAIEAssignBufferAddresses.cpp

```diff
@@ -46,6 +50,204 @@ static LogicalResult basicAllocation(
+  if (!memBankAttr) return false;
+
+  int mem_bank = memBankAttr.getInt();
+  int64_t startAddr = nextAddrInBanks[mem_bank];
+  int64_t endAddr = startAddr + getAllocationSize(buffer);
+  if (endAddr > bankLimits[mem_bank].endAddr)
+    return buffer->emitOpError("would override existing mem_bank");
+  setAndUpdateAddressInBank(buffer, startAddr, endAddr, nextAddrInBanks);
+  return true;
+}
+
+// Function that given a buffer will iterate over all the memory banks starting
+// from the given index to try and find a bank with enough space. If it does,
+// it will set the buffer's address and mem_bank attributes and update the
+// nextAddrInBanks vector. If it does not find one with enough space, it will
+// throw an error. Returns true if the buffer was successfully allocated, false
+// otherwise.
+bool setBufferAddress(BufferOp buffer, int numBanks, int &startBankIndex,
+                      std::vector<int64_t> &nextAddrInBanks,
+                      std::vector<BankLimits> &bankLimits) {
+  assert(startBankIndex < numBanks &&
+         "Unexpected input value for startBankIndex");
+  int bankIndex = startBankIndex;
+  bool allocated = false;
+  for (int i = 0; i < numBanks; i++) {
+    int64_t startAddr = nextAddrInBanks[bankIndex];
+    int64_t endAddr = startAddr + getAllocationSize(buffer);
+    if (endAddr <= bankLimits[bankIndex].endAddr) {
+      buffer.setMemBank(bankIndex);
+      setAndUpdateAddressInBank(buffer, startAddr, endAddr, nextAddrInBanks);
+      allocated = true;
+      bankIndex = (bankIndex + 1) % numBanks;
+      startBankIndex = bankIndex;
+      break;
+    }
+    // Move to the next bank
+    bankIndex = (bankIndex + 1) % numBanks;
+  }
+  // If no bank has enough space, throws an error.
+  if (!allocated) {
+    buffer.emitError("Failed to allocate buffer: ")
+        << buffer.name() << " with size: " << getAllocationSize(buffer)
+        << " bytes on any of the bank.";
+    return false;
+  }
+  return true;
+}
+
+// Function to deallocate attributes of buffers in case of a failure.
+void deAllocateBuffers(SmallVector<BufferOp> &buffers) {
+  for (BufferOp buffer : buffers) {
+    buffer->removeAttr("address");
+    buffer->removeAttr("mem_bank");
+  }
+}
+
+LogicalResult bankAwareAllocation(
+    DenseMap<TileOp, SetVector<BufferOp>> &tileToBuffers,
+    AMDAIEDeviceModel deviceModel) {
+  for (auto &&[tile, buffers] : tileToBuffers) {
+    int maxDataMemorySize;
+    if (deviceModel.isMemTile(tile.getCol(), tile.getRow())) {
+      maxDataMemorySize =
+          deviceModel.getMemTileSize(tile.getCol(), tile.getRow());
+    } else {
+      maxDataMemorySize =
+          deviceModel.getLocalMemorySize(tile.getCol(), tile.getRow());
+    }
+
+    int numBanks = deviceModel.getNumBanks(tile.getCol(), tile.getRow());
+    int bankSize = maxDataMemorySize / numBanks;
+
+    // Each entry of `nextAddrInBanks` is the next address available for use
```

Comment:
Is this code duplicated with the linear algorithm?

Reply:
The sequential scheme doesn't have the `mem_bank` concept and doesn't have each buffer assigned to a particular bank. So I don't see the code duplicated.

Reply:
Or maybe you mean the codes to get `maxDataMemorySize` for each tile? That part is duplicated I think.